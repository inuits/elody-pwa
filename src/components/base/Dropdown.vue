<template>
  <label class="block w-full">
    <span
      v-if="label"
      :class="[`ml-1 text-${labelColor} text-sm ${isDisabled ? 'opacity-50' : ''}`]"
      >{{ label }}</span
    >
    <select
      v-model="selectedItem"
      :class="isDisabled ? 'opacity-40 cursor-not-allowed' : ''"
      :disabled="isDisabled"
    >
      <option v-for="option in options" :key="option" :value="option">
        {{ option }}
      </option>
    </select>
  </label>
</template>

<script lang="ts">
  import { defineComponent, ref, watch } from 'vue';
  import { Unicons } from '@/types';

  export default defineComponent({
    name: 'Dropdown',
    props: {
      label: { type: String, default: '' },
      options: { type: Array, required: true },
      isDisabled: { type: Boolean, required: false },
      modelValue: { type: String, default: undefined },
      bgColor: {
        type: String,
        default: 'neutral-0',
      },
      labelColor: {
        type: String,
        default: 'neutral-600',
      },
    },
    emits: ['update:modelValue'],
    setup(props, { emit }) {
      const selectedItem = ref(props.modelValue);
      watch(selectedItem, (value) => emit('update:modelValue', value));
      return { Unicons, selectedItem };
    },
  });
</script>

<style lang="postcss" scoped>
  select {
    @apply block mr-4 p-2 w-48 min-w-0 cursor-pointer;
    @apply border border-neutral-30;
    @apply text-neutral-700 text-sm;
    @apply rounded bg-neutral-20;
    @apply focus:outline-none;
  }
  option {
    @apply rounded py-2 px-4 bg-neutral-20 h-9;
  }
</style>
