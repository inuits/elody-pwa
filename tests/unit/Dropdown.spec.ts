import { mount, shallowMount } from '@vue/test-utils';
import Dropdown from '@/components/base/Dropdown.vue';
jest.mock('@/types');

let wrapper: any;

describe('Dropdown.vue', () => {
  beforeEach(() => {
    wrapper = shallowMount(Dropdown, {
      props: {
        options: ['5', '10', '15', '20'],
        modelValue: '5',
      },
    });
  });

  it('Renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Option gets selected value for options', () => {
    expect(wrapper.vm.selectedItem).toMatch('5');
  });

  it('renders the correct label', () => {
    const label = wrapper.find('label');
    expect(label.text()).toMatch('');
  });

  it('Updated reactive selected value', () => {
    const option = wrapper.find('option');
    option.setValue('10');
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.selectedItem.value).toBe('10');
      expect(wrapper.emitted('update:selected')).toBe('10');
    });
  });
});
