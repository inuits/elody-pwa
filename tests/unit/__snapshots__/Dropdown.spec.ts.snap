// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown.vue Renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <label
    class="block text-sm text-gray-700 ml-3"
  />
  <div
    class="block flex justify-center ml-3"
  >
    <select
      class="w-48 rounded h-12"
    >
      
      <option
        value="option"
      >
        5
      </option>
      <option
        value="option"
      >
        10
      </option>
      <option
        value="option"
      >
        15
      </option>
      <option
        value="option"
      >
        20
      </option>
      
    </select>
  </div>
  
</div>
`;
